CPP => C Scheduler

ConnorDoyle opened the pull request:
See: http://doc.crates.io/build-script.html
cc @spacejam @jsancio

ConnorDoyle committed:
Added smoke test binary.

ConnorDoyle committed:
First pass delegating from CPP scheduler to C functions.

ConnorDoyle committed:
Fixed gcc compiler errors.

jsancio commented on build.rs:
```diff
@@ -0,0 +1,10 @@
+        "resources/mesos-c/scheduler_driver.cpp",
+    ];
+
+    gcc::compile_library("libmesosc.a", sources);
```
Do we need to tell gcc to "enable" c++ compiling. I think that turn on a bunch of flags like libstdc++ linking.

ConnorDoyle replied:
Good point, enabled.

ConnorDoyle committed:
Enable cpp compilation in build.rs.

ConnorDoyle committed:
Added instructions to link libmesos, libprotobuf-lite.

ConnorDoyle commented:
Got things linking with additional flags to `rustc` (see updated README in this branch).  Seeing a new issue related to protobuf deserialization but this looks like real progress :)

```
# connor at mabuya in ~/projects/ConnorDoyle/mesos-rust on git:c-scheduler o [23:35:59]
$ ./target/debug/test_scheduler                                         
framework_info: [user: "root" name: "mesos-rust-test"]
Starting scheduler driver
MesosSchedulerDriver::create_callbacks
[libprotobuf ERROR google/protobuf/message_lite.cc:123] Can't parse message of type "mesos.FrameworkInfo" because it is missing required fields: user, name
Deserialization failed
```

ConnorDoyle committed:
Added ProbufObj <=> protobuf::Message conversion tests.

ConnorDoyle committed:
Fix lifetime of backing vector in ProtobufObj::from_message.

ConnorDoyle commented:
The backing vector was being freed before deserialization.  Updated the signature of `ProtobufObj::from_message` to take a mutable reference to storage to allow the caller to manage the lifetime of this object.  May revisit that API in the future to clean it up.

ConnorDoyle committed:
Added logic to start and join the scheduler driver in run().

ConnorDoyle commented:
Looks like we're in business!

```
$ ./target/debug/test_scheduler                                                                       
framework_info: [user: "root" name: "mesos-rust-test"]
MesosSchedulerDriver::create_callbacks
WARNING: Logging before InitGoogleLogging() is written to STDERR
W0828 00:35:04.814226 2028991248 sched.cpp:1326] 
**************************************************
Scheduler driver bound to loopback interface! Cannot communicate with remote master(s). You might want to set 'LIBPROCESS_IP' environment variable to use a routable IP address.
**************************************************
Starting scheduler driver
I0828 00:35:04.821302 2028991248 sched.cpp:157] Version: 0.23.0
scheduler_status: [2]
Joining scheduler driver
I0828 00:35:04.823639 119578624 sched.cpp:254] New master detected at master@127.0.0.1:5050
I0828 00:35:04.823827 119578624 sched.cpp:264] No credentials provided. Attempting to register without authentication
I0828 00:35:04.825016 119578624 sched.cpp:448] Framework registered with 20150828-003422-16777343-5050-12379-0000
wrapped_registered_callback
delegating to scheduler.registered...
```

ConnorDoyle committed:
Delegate cpp scheduler registered callback to Rust

ConnorDoyle commented:
Hold on to your butts...

```
connor@iota:mesos-rust (c-scheduler) $ ./test_scheduler 
framework_info: [user: "root" name: "mesos-rust-test"]
MesosSchedulerDriver::create_callbacks
WARNING: Logging before InitGoogleLogging() is written to STDERR
W0828 19:14:00.197418 1915597568 sched.cpp:1487] 
**************************************************
Scheduler driver bound to loopback interface! Cannot communicate with remote master(s). You might want to set 'LIBPROCESS_IP' environment variable to use a routable IP address.
**************************************************
Starting scheduler driver
I0828 19:14:00.203116 1915597568 sched.cpp:164] Version: 0.25.0
scheduler_status: [2]
Joining scheduler driver
I0828 19:14:00.206955 211447808 sched.cpp:262] New master detected at master@127.0.0.1:5050
I0828 19:14:00.207201 211447808 sched.cpp:272] No credentials provided. Attempting to register without authentication
I0828 19:14:00.208397 211447808 sched.cpp:640] Framework registered with 20150828-191355-16777343-5050-65961-0000
wrapped_registered_callback
delegating to scheduler.registered...
MyScheduler::registered
framework_id: value: "20150828-191355-16777343-5050-65961-0000"
master_info: id: "20150828-191355-16777343-5050-65961" ip: 16777343 port: 5050 pid: "master@127.0.0.1:5050" hostname: "localhost" version: "0.25.0"
```

ConnorDoyle committed:
Wrapped low-level delegation in native/mod.rs.

- Added support for reregister callback in Scheduler.
- Added support for resourceOffers callback in Scheduler.

ConnorDoyle commented:
Added support for `reregister` and `resourceOffers`.

```
connor@mabuya:mesos-rust (c-scheduler) $ ./test_scheduler                                connor@mabuya:mesos-rust (c-scheduler) $ ./test_scheduler
framework_info: [user: "root" name: "mesos-rust-test"]
WARNING: Logging before InitGoogleLogging() is written to STDERR
W0829 01:28:16.220099 2028991248 sched.cpp:1326] 
**************************************************
Scheduler driver bound to loopback interface! Cannot communicate with remote master(s). You might want to set 'LIBPROCESS_IP' environment variable to use a routable IP address.
**************************************************
Starting scheduler driver
I0829 01:28:16.227249 2028991248 sched.cpp:157] Version: 0.23.0
scheduler_status: [2]
Joining scheduler driver
I0829 01:28:16.229468 284360704 sched.cpp:254] New master detected at master@127.0.0.1:5050
I0829 01:28:16.229660 284360704 sched.cpp:264] No credentials provided. Attempting to register without authentication
I0829 01:28:16.230854 284360704 sched.cpp:448] Framework registered with 20150829-012812-16777343-5050-15835-0000
MyScheduler::registered
framework_id: value: "20150829-012812-16777343-5050-15835-0000"
master_info: id: "20150829-012812-16777343-5050-15835" ip: 16777343 port: 5050 pid: "master@127.0.0.1:5050" hostname: "localhost" version: "0.23.0"
MyScheduler::resource_offers
Received [1] offers
  Offer id: [value: "20150829-012812-16777343-5050-15835-O0"]
```

There might be a bug related to how the Offer protobufs are being deserialized however, or in the generated Rust proto code itself.  Ran into some `None.unwrap()` errors when trying to print the resulting vector.

bonifaido commented:
Printing messages with rust keywords as field names has some issues in rust-protobuf, see:
https://github.com/stepancheg/rust-protobuf/issues/105

ConnorDoyle commented:
Thanks for the clue @bonifaido!

ConnorDoyle committed:
Renamed  protobuf fields to workaround mesos-protobuf bug.

ConnorDoyle committed:
Added basic Makefile.

ConnorDoyle committed:
Fixed string allocation in C scheduler.

ConnorDoyle commented:
Merging this since it seems stable though the three callbacks I've wired up so far.  Let's add the rest (and the driver calls) in subsequent patches :+1:
